### ISSUE TITLE
[css-position-3] Reinterpret viewport positioned (fixed, sticky) elements wrt virtual keyboard

### ISSUE THREAD

**Issue by flackr** [w3c/csswg-drafts#7475]
Position [fixed](https://www.w3.org/TR/css-position-3/#valdef-position-fixed) is positioned and sized relative to a [fixed position containing block](https://www.w3.org/TR/css-position-3/#fixed-positioning-containing-block) which in the absence of no established container is the [viewport](https://www.w3.org/TR/CSS2/visuren.html#x1).

The behavior today when a virtual keyboard is shown varies (see [demo](https://flackr.github.io/web-demos/viewport/) for interactive examples). In each screenshot, the left image is what the user would see, and the right is a visualization of the viewports with the blue outline being the size of the ICB / layout viewport / 100vh and the red outline being the visual viewport:
1. [iOS (and ChromeOS)](https://flackr.github.io/web-demos/viewport/#behavior=ios) shrink the visual viewport but keep the fixed position viewport the same size. As a result, elements which are fixed position to the top/bottom may be scrolled out of view (even while zoomed out). However, showing the keyboard does not affect layout in any way.

   ![ios-chromeos](https://user-images.githubusercontent.com/366761/177874302-0d7ec7b4-9624-4083-b31b-ee3110b9f04d.jpg)

2. [Android](https://flackr.github.io/web-demos/viewport/#behavior=android) resizes the page to the height remaining after the keyboard is shown, changing the fixed position viewport as well as the ICB. This has the consequence of doing a relayout and potentially triggering different media query breakpoints, but also means that position: fixed elements are still fully visible while zoomed out.

   ![android](https://user-images.githubusercontent.com/366761/177874572-c54e436d-fd6f-409e-90f0-7af3726cfdea.jpg)

We would like to unify the behaviors, and agree that not having a global resize is better, however one of the stumbling blocks when we tried to adopt the iOS model is that developers struggle with being able to create bottom fixed position elements, which has lead to various alternate proposals on top of the iOS model. Let's explore some:

3. [device-fixed](https://flackr.github.io/web-demos/viewport/#behavior=device-fixed) is a [proposed position]((https://www.quirksmode.org/blog/archives/2010/12/the_fifth_posit.html#:~:text=the%20user%20interaction.-,position%3A%20device%2Dfixed,-Which%20of%20these)) to position elements to the fixed viewport regardless of zoom. This fixes the issue when the keyboard is shown as the element is shifted up, however when zooming `position: device-fixed` content remains the same size which raises concerns for accessibility as well as not preserving the relative size of content on the page.

   ![device-fixed](https://user-images.githubusercontent.com/366761/177882582-ab5c96b3-c9fa-4d6a-ab8b-59ad012ea8b9.jpg)

4. [Scaled device-fixed](https://flackr.github.io/web-demos/viewport/#behavior=device-fixed-scaled) is an alternative which preserves the zoom scale but still treats the visual viewport rect as the fixed position rect. This allows you to zoom in on `position: device-fixed` content but this results in the content intruding more into the viewport.

   ![scaled-device-fixed](https://user-images.githubusercontent.com/366761/177882784-2e081053-70cd-4ec4-a031-2e1876c1f0c1.jpg)

5. I propose [excluding the keyboard from the position fixed viewport](https://flackr.github.io/web-demos/viewport/#behavior=fixed-viewport). This leads to a very similar behavior to the android one in that `position: fixed` content avoids the keyboard, except that it would not change the height of the ICB or viewport units. The only content which needs to relayout is that which is sized with respect to the fixed position viewport, and content which is not sized with respect to the viewport can be shifted on the compositor.

   ![fixed-viewport](https://user-images.githubusercontent.com/366761/177883064-f6bf1e1b-cb69-4544-98fb-1e43f513f662.jpg)

   When you zoom in the same thing would happen as currently happens on android, the visual viewport would zoom in but leave the fixed position viewport size unchanged.

   ![fixed-viewport-zoomed](https://user-images.githubusercontent.com/366761/177883148-f081ef40-fe01-4e10-b3fb-b8ff85e67ae5.jpg)

   Another advantage of this option is that we could easily use the same adjusted fixed viewport as the scrollport for sticky position elements on the root, whereas options 3 and 4 would not implicitly add any support for sticky without adding e.g. device-sticky.

TLDR; In order to close the interop gap while still allowing developers to position fixed position content above the keyboard I propose we adopt option 5.

Each option links to a demo of that option which can be played with on a desktop browser.

**Comment by bramus** [comment]
I’ve given all proposals in here a lot of thought and while I do like the general idea behind Option 5 _([PosFixedViewport](https://flackr.github.io/web-demos/viewport/#behavior=fixed-viewport))_. But, I have a few concerns to be fully convinced of it:

1. This would alter the existing behavior of `position: fixed` on all iOS browsers + Chrome on CrOS. [This demo here](https://cdpn.io/pen/debug/eYMZEaV/addc16387bdd13083e05948afcbb8d63) – which mimics a bottom navigation bar – for example would no longer work as the author expected it to work. This will annoy authors.
1. Authors cannot easily feature detect wether the new or old behavior is being used. Therefore they cannot predict how the site will look with the OSK shown. This will annoy authors as well.
1. This does not provide a solution for authors to stick something against the Visual Viewport, something that might come in handy for Floating Action Buttons or Toolbars which do need to stay above the zoomed content.

Could we solve this differently? Instead of overloading the position property with new values – such as `device-fixed`, `visual-viewport-fixed`, … – could we extend its syntax and provide the viewport as an extra argument? I’m thinking of:

- `position: fixed / layout` = current behavior, would be the same as `position: fixed`)
- `position: fixed / visual` = fixed against the visual viewport, also when zoomed in
- `position: fixed / fixed` _(* lacking a better name)_ = the proposed behavior from Option 5 _([PosFixedViewport](https://flackr.github.io/web-demos/viewport/#behavior=fixed-viewport))_

This way:

1. The behavior of `position: fixed` would not be changed, making the adjustments backwards compatible
2. Authors can feature detect the new behavior and act upon that
3. Authors can also stick something against the Visual Viewport

Either way, I think we should definitely adjust Chrome on Android to no longer resize the ICB when the OSK gets shown. This can be implemented separately from the outcome here.

**Comment by flackr** [comment]
> 1. This would alter the existing behavior of `position: fixed` on all iOS browsers + Chrome on CrOS. [This demo here](https://cdpn.io/pen/debug/eYMZEaV/addc16387bdd13083e05948afcbb8d63) – which mimics a bottom navigation bar – for example would no longer work as the author expected it to work. This will annoy authors.

You proposed no longer resizing the ICB on Android (without this reinterpretation), which would have the opposite effect of annoying authors who are currently expecting the bottom bar to stay visible wouldn't it? We will be annoying some group of users when we align the keyboard behavior, right? Do you think that users who wanted the bottom / top bar sometimes obscured by the keyboard will be greater in number / more annoyed than the users who didn't? My expectation was that more devs would have expected fixed pos elements to stay visible (the Android behavior) and that changing the default fixed behavior to no longer do that would be more annoying.

Note that a developer could build the old behavior with `navigator.virtualKeyboard.overlaysContent = true` when iOS adds support for this.

> 2. Authors cannot easily feature detect wether the new or old behavior is being used. Therefore they cannot predict how the site will look with the OSK shown. This will annoy authors as well.

If we launch this change in coordination with no longer resizing the ICB, there will generally be no visual change for developers. Whereas if we don't, there will be. This does risk annoying developers on iOS / ChromeOS but they would also currently be annoyed by the Android behavior. The `navigator.virtualKeyboard.overlaysContent` allows developers to design their own consistent behavior though it is admittedly not yet supported on iOS.

> 3. This does not provide a solution for authors to stick something against the Visual Viewport, something that might come in handy for Floating Action Buttons or Toolbars which do need to stay above the zoomed content.

Presumably this is the device-fixed (option 3) in the above list, right? I.e. not changing with zoom. I do think that this is valuable. My main concern is that device-fixed is not the right behavior for bottom-bars, so even if we add it, it does not solve the bottom-bar issue. I do think `device-fixed` may be independently valuable even though it does have accessibility concerns.

> Could we solve this differently? Instead of overloading the position property with new values – such as `device-fixed`, `visual-viewport-fixed`, … – could we extend its syntax and provide the viewport as an extra argument? I’m thinking of:
> 
> * `position: fixed / layout` = current behavior, would be the same as `position: fixed`)
> * `position: fixed / visual` = fixed against the visual viewport, also when zoomed in

To be clear, is this option 3 or option 4?

> * `position: fixed / fixed` _(* lacking a better name)_ = the proposed behavior from Option 5 _([PosFixedViewport](https://flackr.github.io/web-demos/viewport/#behavior=fixed-viewport))_

> 1. The behavior of `position: fixed` would not be changed, making the adjustments backwards compatible

We will be changing the position: fixed behavior with the no longer resizing the ICB change, which is not backwards compatible on Android without an update to their behavior.

> 2. Authors can feature detect the new behavior and act upon that

You could feature detect this when the keyboard is shown by comparing the height of 100% fixed to the viewport. Alternately the virtualKeyboard API lets developers opt into alternate treatments.

> 3. Authors can also stick something against the Visual Viewport.

Agreed, this is an independently useful feature - but not the best for bottom-bars.

**Comment by fantasai** [comment]
There's some useful discussion of use cases in https://github.com/w3c/csswg-drafts/issues/7194 btw.

**Comment by bramus** [comment]
Had a breakout session with @smfr at the 2023 Cupertino F2F. In summary, two concerns were raised:

- Landscape Mode: there is almost no room left above the keyboard, so we should be cautious here.
- In the scenario where a user first pinch-zooms and then bring up keyboard, a FixedPos element can suddenly jump into view because the PosFixedViewport shrinks.

Furthermore, the default behavior we have right now should preferably not be changed, and it should be an opt-in indeed.

**Comment by flackr** [comment]
> Furthermore, the default behavior we have right now should preferably not be changed, and it should be an opt-in indeed.

We currently don't have interop though right? I think one of the current behaviors should change.

**Comment by una** [comment]
I like this idea, being able to specify *where* something is fixed, but I would propose a function syntax, like `fixed(visual-viewport)`.  This would make it more clear that this is an argument for where you are applying the positioning algorithm. The`/` syntax reminds me of positioning that breaks out two directions (i.e. block and inline as in grid). It would also be neat to be able to reference names areas here! I.e. `fixed(--area)`

**Comment by bramus** [comment]
While Visual Viewport units have already been rejected – see https://github.com/w3c/csswg-drafts/issues/7194 – the reinterpretation/extension to position fixed from this issue could help out here.

You could bottom anchor the one thing the layout viewport and the other to the (unzoomed) visual viewport.

**Comment by ryantownsend** [comment]
@bramus `position: fixed / [fixed/visual]` seems sensible, but to avoid dependence on knowing the height of elements in my example, it would be useful to extend this option by also introducing `position: sticky / [fixed/visual]`, otherwise we'd be forever needing to bottom-pad the container with the height of the floating element – which is doable but far less maintainable/responsive.

**Comment by ryantownsend** [comment]
Having now read more of the above thread, has a ‘zoomed’ media query been suggested/considered?

If there are concerns with an FAB dominating the viewport when zoomed or performance being impacted when evaluating viewport units during zoom, maybe a boolean-, threshold- or step-based media query could be a compromise?

e.g. `@media (1 < zoom-level < 2)` (greater than 100%, less than 200%), akin to [the level 4 width queries](https://web.dev/articles/media-query-range-syntax)

`@media (zoom-threshold: 2)` – more basic threshold

`@media (zoomed: true)` - most basic boolean.

(Note: I’m suggesting only one of these would be implemented, they just have varying degrees of complexity)

This would present a multitude of options to authors, e.g. with any of the above, we could opt to only pin while the viewport remains unzoomed:

```css
@supports (position: fixed / fixed) {
  @media not (zoomed: true) {
    .fab { position: fixed / fixed }
  }
}
```

With `zoom-level`, we could adjust sizes for different zoom levels:

```css
@supports (position: fixed / fixed) {
  .fab { position: fixed / fixed }

  @media (zoom-level > 1) {
    .fab { font-size: 0.8em }
  }

  @media (zoom-level > 2) {
    .fab { font-size: 0.6em }
  }
}
```

**Comment by waterplea** [comment]
It's pretty crazy that in 2024 we still cannot show an always visible floating action button at the bottom of the screen with CSS, regardless of screen keyboard being present or not. That's a very needed feature.
